src/PartitionLoops.cpp

```diff
@@ -17,748 +19,652 @@ using std::string;
-            // conditions, which means the prologue is the union of
-            // multiple negated conditions, so any individual one of
-            // them might not apply.
-            return substitute(prologue_replacements[0].old_expr,
-                              prologue_replacements[0].new_expr,
-                              s);
+    void visit(const Store *op) {
+        bool old_in_index = in_index;
+        in_index = true;
+        Expr index = mutate(op->index);
+        in_index = old_in_index;
+        Expr value = mutate(op->value);
+        if (index.same_as(op->index) && value.same_as(op->value)) {
+            stmt = op;
         } else {
-            return s;
+            stmt = Store::make(op->name, value, index);
         }
     }
 
-    Stmt simplify_epilogue(Stmt s) {
-        if (max_vals.size() == 1 &&
-            epilogue_replacements.size() == 1) {
-            return substitute(epilogue_replacements[0].old_expr,
-                              epilogue_replacements[0].new_expr,
-                              s);
+    bool in_index = false;
+};
+
+// Remove any 'likely' intrinsics.
+class RemoveLikelyTags : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const Call *op) {
+        if (op->name == Call::likely && op->call_type == Call::Intrinsic) {
+            internal_assert(op->args.size() == 1);
+            expr = mutate(op->args[0]);
         } else {
-            return s;
+            IRMutator::visit(op);
         }
     }
+};
 
-    /** Wrap a statement in any common subexpressions used by the
-     * minvals or maxvals. */
-    Stmt add_containing_lets(Stmt s) {
-        for (size_t i = containing_lets.size(); i > 0; i--) {
-            const string &name = containing_lets[i-1].first;
-            Expr value = containing_lets[i-1].second;
-
-            // Subexpressions are commonly shared between minval
-            // expressions and the maxval expressions.
-            for (size_t j = 0; j < i-1; j++) {
-                // Just refer to the other let.
-                if (equal(containing_lets[j].second, value)) {
-                    value = Variable::make(value.type(), containing_lets[j].first);
-                    break;
-                }
-            }
-
-            s = LetStmt::make(name, value, s);
+class HasLikelyTag : public IRVisitor {
+    using IRVisitor::visit;
+    void visit(const Call *op) {
+        if (op->name == Call::likely &&
+            op->call_type == Call::Intrinsic) {
+            result = true;
+        } else {
+            IRVisitor::visit(op);
         }
-        return s;
     }
+public:
+    bool result = false;
+};
 
-private:
-
-    vector<Expr> min_vals, max_vals;
-
-    string loop_var;
-    Scope<Expr> bound_vars;
-    Scope<int> free_vars;
-    Scope<int> inner_loop_vars;
+bool has_likely_tag(Expr e) {
+    HasLikelyTag h;
+    e.accept(&h);
+    return h.result;
+}
 
+struct Simplification {
```

Comment:
How about a comment here explaining the motivation for finding simplifications on subsets of the domain for a loop variable? I think it would help to understand what this is doing when you later understand that the steady state is taken to be the intersection of all of the subsets, where all of the simplifications can be applied.